lib/protobuf/json/encode.ex

```diff
@@ -0,0 +1,139 @@
+defmodule Protobuf.JSON.Encode do
+  @moduledoc """
+  Internal JSON encoding API.
+  """
+
+  alias Protobuf.MessageProps
+
+  @type encode_opt ::
+          {:use_proto_names, boolean}
+          | {:use_enum_numbers, boolean}
+          | {:emit_unpopulated, boolean}
+
+  @compile {:inline,
+            encode_field: 3,
+            encode_key: 2,
+            maybe_repeat: 3,
+            encode_float: 1,
+            encode_enum: 3,
+            safe_enum_key: 2}
+
+  @doc false
+  @spec encode(struct, MessageProps.t(), [encode_opt]) :: map
+  def encode(%_{} = struct, %MessageProps{} = message_props, opts \\ []) do
+    regular = encode_regular_fields(struct, message_props, opts)
+    oneofs = encode_oneof_fields(struct, message_props, opts)
+
+    :maps.from_list(regular ++ oneofs)
+  end
+
+  defp encode_regular_fields(struct, %{field_props: field_props}, opts) do
+    for {_field_num, %{name_atom: name, oneof: nil} = prop} <- field_props,
+        %{^name => value} = struct,
+        opts[:emit_unpopulated] || !default?(prop, value) do
+      encode_field(prop, value, opts)
+    end
+  end
+
+  defp encode_oneof_fields(struct, message_props, opts) do
+    %{field_tags: field_tags, field_props: field_props, oneof: oneofs} = message_props
+
+    for {oneof_name, _index} <- oneofs,
+        tag_and_value = Map.get(struct, oneof_name) do
+      {tag, value} = tag_and_value
+      prop = field_props[field_tags[tag]]
+      encode_field(prop, value, opts)
+    end
+  end
+
+  # TODO: handle invalid values? check types?
+  defp encode_field(prop, value, opts) do
+    {encode_key(prop, opts), encode_value(value, prop, opts)}
+  end
+
+  defp encode_key(prop, opts) do
+    if opts[:use_proto_names], do: prop.name, else: prop.json_name
+  end
+
+  @int32_types ~w(int32 sint32 sfixed32 fixed32 uint32)a
+  @int64_types ~w(int64 sint64 sfixed64 fixed64 uint64)a
+  @float_types [:float, :double]
+  @raw_types [:string, :bool] ++ @int32_types
+
+  defp encode_value(nil, _prop, _opts), do: nil
+
+  defp encode_value(value, %{type: type} = prop, _opts) when type in @raw_types do
+    maybe_repeat(prop, value, & &1)
+  end
+
+  defp encode_value(value, %{type: type} = prop, _opts) when type in @int64_types do
+    maybe_repeat(prop, value, &Integer.to_string/1)
+  end
+
+  defp encode_value(value, %{type: :bytes} = prop, _opts) do
+    maybe_repeat(prop, value, &Base.encode64/1)
```

Comment:
Do we care in any way (in JSON, in general) if this base64 is lowercase or uppercase?

Reply:
Good question. I'll double-check that.

Reply:
Base64 itself is case-sensitive, so it is important to respect that. Map keys can be `bytes`, but they will always be encoded/decoded according to the rules for `bytes`, so it won't be a problem.

Reply:
Great 👍